### PR TITLE
Support auto-splatting in captured block literals

### LINKED ISSUE #7706: Iterable of tuple, methods select/reject/map do not unpack tuple
Very similar to #3419, which was specific to the sort_by method. There are many other methods which also cannot unpack tuples automatically. Eg.

```crystal
[{1,2}].each.select { |x,y| y > x }
```
will cause a compile error.

Note for users looking at this bug you can work around this limitation by manually unpacking the args like so:

```crystal
[{1,2}].each.select { |(x, y)| y > x }
```

I don't think this necessarily has to be fixed right this instant, but this is IMO a big usability bug. There's no clear way to inspect the type signature of a method to tell if it can unpack tuples for you. There's no clear pattern to why some methods will unpack tuples and why some won't. This is not included in the documentation. Finally, the error `Error in line 1: wrong number of block arguments (given 2, expected 1)` is kind of confusing because the block would be called with 1 argument, while it expects 2.

Perhaps this would be a better error message:  `Error in line 1: wrong number of block arguments (given block accepting 2 arguments, expected block accepting 1)`.

Edited: Originally the code examples above used `[{1, 2}].select` without the intermediate `each`, meaning they were calling the `select` method defined by Enumerable, which is not affected by this issue.

### PR DESCRIPTION
**Pull request by HertzDevil**
Resolves #7706.

In a case like

```crystal
def foo(&block : {Int32, Char} -> _)
  tup = {1, 'a'}
  block.call tup
end

foo do |x, y|
  {x, y}
end
```

The compiler now rewrites the block into the following `ProcLiteral`:

```crystal
->(__temp : {Int32, Char}) do
  x = __temp[0]
  y = __temp[1]
  {x, y}
end
```

The conditions for auto-splatting are the same as non-captured blocks: the block must accept a single `Tuple` instance, there must be 2 or more block parameters, and none of these parameters must be a splat.

When there are fewer block parameters than tuple elements, the whole tuple is still passed but the `__temp` assignments corresponding to the missing parameters will be gone.

Tuple unpacking of individual block parameters is also possible. For example, given

```crystal
def foo(&block : Tuple(Tuple(Int32, Char), Bool, Float64) -> _)
  tup = { {1, 'a'}, true, 1.0 }
  block.call tup
end

foo do |(x, y), z|
  {x, y, z} # => {1, 'a', true}
end
```

The expanded `Proc` is more or less equivalent to

```crystal
->(__temp_1 : Tuple(Tuple(Int32, Char), Bool, Float64)) do
  __arg0 = __temp_1[0]
  z = __temp_1[1]
  # __temp_1[2] not accessed

  x = __arg0[0]
  y = __arg0[1]

  {x, y, z}
end
```

Internally `(x, y)` is transformed into `__arg0` by the _parser_; `__temp_1` is injected in the semantic phase when the block is actually matched against the enclosing method.

### CODE REVIEW

**Review by asterite — APPROVED** [review]
Awesome!

File: src/compiler/crystal/semantic/call.cr [hunk 1]
```diff
@@ -790,18 +790,22 @@ class Crystal::Call
...
       # Check if tuple unpacking is needed
-      if yield_vars.size == 1 &&
-         (yield_var_type = yield_vars.first.type).is_a?(TupleInstanceType) &&
-         block.args.size > 1
+      if auto_unpack_needed = (
+           yield_vars.size == 1 &&
+           (yield_var_type = yield_vars.first.type.as?(TupleInstanceType)) &&
+           block.args.size > 1 &&
+           !block.splat_index
+         )
```
Comment (straight-shoota) [inline src/compiler/crystal/semantic/call.cr:802]: Stylistically, I would suggest to move such complex assignments out of the condition to a separate expression.

**Review by straight-shoota — APPROVED** [review]

**Comment by bcardiff** [comment]
FYI, This CI is complaining for real in this PR in spec/compiler/semantic/closure_spec.cr:385, although for some reason it didn't fail on all platforms 🤔

**Comment by HertzDevil** [comment]
Ah this is because I forgot to change some expected `assert_error` messages. Push will be up soon...

**Comment by bcardiff** [comment]
It's still failing just on some platforms. The merge from master was done in an old commit. I am not sure if that's the reason.

**Review by straight-shoota — APPROVED** [review]

**Pull request merged**